src/c/coap-server.c

```diff
@@ -0,0 +1,465 @@
+  uint8_t data_str[INT32_STR_MAXLEN+1];
+  memset (data_str, 0, INT32_STR_MAXLEN+1);
+  memcpy (data_str, data, len);
+
+  char *endptr;
+  errno = 0;
+  long int_val = strtol ((char *)data_str, &endptr, 10);
+
+  /* validate strtol conversion, and ensure within range */
+  if (errno || (*endptr != '\0') || (int_val < INT32_MIN) || (int_val > INT32_MAX))
+  {
+    iot_log_info (sdk_ctx->lc, "invalid int32 of len %u", len);
+    return NULL;
+  }
+
+  return iot_data_alloc_i32 ((int32_t) int_val);
+}
+
+/* Caller must free returned iot_data_t */
+static iot_data_t*
+read_data_string (uint8_t *data, size_t len)
+{
+  /* must copy request data to append null terminator */
+  char *str_data = malloc (len + 1);
+  memcpy (str_data, data, len);
+  str_data[len] = '\0';
+
+  iot_data_t *iot_data = iot_data_alloc_string(str_data, IOT_DATA_COPY);
+  free (str_data);
+
+  return iot_data;
+}
+
+/*
+ * Parse URI path, expect 3 segments: /a1r/{device-name}/{resource-name}
+ *
+ * @param[in] request For path to parse
+ * @param[out] device Found device
+ * @param[out] resource Found resource for device
+ * @return true if URI format OK, and device and resource found 
+ */
+static bool
+parse_path (coap_pdu_t *request, edgex_device **device_ptr, edgex_deviceresource **resource_ptr)
+{
+  coap_string_t *uri_path = coap_get_uri_path (request);
+  iot_log_debug (sdk_ctx->lc, "URI %s", uri_path->s);
+  char *path = (char *)uri_path->s;
+
+  edgex_device *device = NULL;
+  edgex_deviceprofile *profile = NULL;
+  edgex_deviceresource *resource = NULL;
+  
+  char *seg = strtok (path, "/");
+  bool res = false;
+  for (int i = 0; i < 3; i++)
+  {
+    if (!seg)
+    {
+      iot_log_info (sdk_ctx->lc, "missing URI segment %u", i);
+      break;
+    }
+
+    switch (i)
+    {
+    case 0:
+      if (strcmp (seg, RESOURCE_SEG1))
+      {
+        iot_log_info (sdk_ctx->lc, "invalid URI; segment %u", i);
+        goto end_for;
+      }
+      break;
+    case 1:
+      if (!(device = edgex_get_device_byname (sdk_ctx->service, seg)))
+      {
+        iot_log_info (sdk_ctx->lc, "device not found: %s", seg);
+        goto end_for;
+      }
+      profile = device->profile;
+      break;
+    case 2:
+      for (; profile; profile = profile->next)
+      {
+        resource = profile->device_resources;
+        for (; resource; resource = resource->next)
+        {
+          if (!strcmp (resource->name, seg))
+          {
+            break;
+          }
+        }
+      }
+      if (!resource)
+      {
+        iot_log_info (sdk_ctx->lc, "resource not found: %s", seg);
+        goto end_for;
+      }
+      res = true;
+      break;
+    }
+    seg = strtok (NULL, "/");
+  }
+
+ end_for:
+  if (res && seg)
+  {
+    iot_log_info (sdk_ctx->lc, "extra URI segment");
+    res = false;
+  }
+  coap_delete_string (uri_path);
+
+  if (res)
+  {
+    *device_ptr = device;
+    *resource_ptr = resource;
+  }
+  else if (device)
```

Comment:
`edgex_free_device()` is like `free()` in that it is ok to call it with a null pointer

Reply:
Thanks. Also found another instance of this use at the end of `data_handler()`.

Addressed in a0106d6